src/System.Net.NetworkInformation/src/System/Net/NetworkInformation/StringParsingHelpers.Addresses.cs

```diff
@@ -31,14 +25,57 @@ internal static List<GatewayIPAddressInformation> ParseGatewayAddressesFromRoute
                     parser.MoveNextOrFail();
                     string gatewayIPHex = parser.MoveAndExtractNext();
                     long addressValue = Convert.ToInt64(gatewayIPHex, 16);
-                    IPAddress address = new IPAddress(addressValue);
-                    collection.Add(new SimpleGatewayIPAddressInformation(address));
+                    if (addressValue != 0)
+                    {
+                        // Skip device routes without valid NextHop IP address.
+                        IPAddress address = new IPAddress(addressValue);
+                        collection.Add(new SimpleGatewayIPAddressInformation(address));
+                    }
                 }
             }
 
             return collection;
         }
 
+        internal static void ParseIPv6GatewayAddressesFromRouteFile(List<GatewayIPAddressInformation> collection, string filePath, string interfaceName, long scopeId)
+        {
+            // Columns are as follows (first-line header):
+            // 00000000000000000000000000000000 00 00000000000000000000000000000000 00 00000000000000000000000000000000 ffffffff 00000001 00000001 00200200 lo
+            // +------------------------------+ ++ +------------------------------+ ++ +------------------------------+ +------+ +------+ +------+ +------+ ++
+            // |                                |  |                                |  |                                |        |        |        |        |
+            // 0                                1  2                                3  4                                5        6        7        8        9
+            //
+            // 0. IPv6 destination network displayed in 32 hexadecimal chars without colons as separator
+            // 1. IPv6 destination prefix length in hexadecimal
+            // 2. IPv6 source network displayed in 32 hexadecimal chars without colons as separator
+            // 3. IPv6 source prefix length in hexadecimal
+            // 4. IPv6 next hop displayed in 32 hexadecimal chars without colons as separator
+            // 5. Metric in hexadecimal
+            // 6. Reference counter
+            // 7. Use counter
+            // 8. Flags
+            // 9. Interface name
+            string[] fileLines = File.ReadAllLines(filePath);
+            foreach (string line in fileLines)
+            {
+                if (line.StartsWith("00000000000000000000000000000000"))
+                {
+                   String[] token = line.Split(null);
+                   if (token.Length > 4 && token[4] != "00000000000000000000000000000000")
+                   {
+
+                        IPAddress address = ParseIPv6HexString(token[4], isNetworkOrder: true);
+                        if (address.IsIPv6LinkLocal)
+                        {
+                            // For Link-Local addresses add ScopeId as that is not part of the route entry.
+                            address.ScopeId = scopeId;
```

Comment:
Should `interfaceName` be used somewhere in this method? Do we need to match against device name?

Reply:
yes, it does. I updated code as well as add more test variants.